HW4_Boggle/src/Boggle.cpp

```diff
@@ -0,0 +1,188 @@
+
+Boggle::Boggle(Lexicon& dictionary, string boardText) : _board(kCubeRow, kCubeColumn), _dictionary(dictionary) {
+    // use the given 16-letter string to initialize the 16 board cubes from top-left to bottom-right.
+    // If the string is empty, generate a random shuffled board.
+    if (boardText.empty()) {
+        boardText = randomInitialize();
+    }
+
+    int k = 0;
+    for (int i = 0; i < kCubeRow; i++) {
+        for (int j = 0; j < kCubeColumn; j++) {
+            _board[i][j] = boardText[k];
+            k++;
+        }
+    }
+}
+
+string Boggle::randomInitialize() {
+    // randomly shuffle the cubes and pick one of the letter facing up.
+    string boardText;
+    shuffle(CUBES, kCubeSize);
+    for (int i = 0; i < kCubeSize; i++) {
+        boardText += CUBES[i][randomInteger(0, kNumFaces-1)];
+    }
+    return boardText;
+}
+
+char Boggle::getLetter(int row, int col) {
+    // return the character that is stored in Boggle board at the given 0-based row and column.
+    // If the row and/or column are out of bounds, throw an int exception.
+    if ((row >= 0) && (row < kCubeRow) && (col >= 0) && (col < kCubeColumn)) {
+        return _board[row][col];
+    }
+    throw out_of_range("Index out of bound.");
+}
+
+bool Boggle::checkWord(string word) {
+    // check whether the given word string is suitable to search for:
+    // that is, whether it is in the dictionary, long enough to be a valid Boggle word,
+    // and has not already been found.
+    // You should return a boolean result of true if the word is suitable, and otherwise you should return false.
+    // method is case-insensitive; you should properly verify the word whether it is passed in upper, lower, or mixed case.
+    return ((word.size() >= 4) && (_dictionary.contains(word)) && (!_validUserInputs.contains(toUpperCase(word))));
+}
+
+Set<string> Boggle::getUserInputs() {
+    return _validUserInputs;
+}
+
+string Boggle::getLastUserInput() {
+    return _lastUserInput;
+}
+
+void Boggle::updateUserStatus(string word) {
+    _validUserInputs.add(toUpperCase(word));
+    _lastUserInput = word;
+    _humanScore += word.length() - 3;
+}
+
+bool Boggle::humanWordSearch(string word) {
+    // perform a search on the board for an individual word.
+    // If the word can be formed, you should add it to the human's set of found words.
+    // function returns a boolean result of whether the word can be formed.
+    // If the word is unsuitable, should not perform the recursive search.
+    word = toUpperCase(word);
+    for (int i = 0; i < kCubeRow; i++) {
+        for (int j = 0; j < kCubeColumn; j++) {
+            Grid<int> visited(kCubeRow, kCubeColumn, 0);
+            visited[i][j] = 1;
+            if (humanWordSearchHelper(i, j, visited, word)) {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+bool Boggle::humanWordSearchHelper(int i, int j, Grid<int>& visited, string word) {

```

Comment:
optimization: do not call word.substring, it will copy the word. can use an integer to record the current index.